ClassCastException due to classloader mismatch with JDK 25

### Describe the bug

With the following scenario:

- Library `libTest` defines a methods for loading classes (basically a wrapper for `Class.forName` and `Thread.currentThread().getContextClassLoader().loadClass`
- The test in an application uses this library to load a class defined in the test sources

Previously this also did not work when using `Class.forName` in the library -- it fails with a `ClassNotFoundException`, howerver, using the class loader of the current thread used to work.

Sadly this seems to break when running with JDK 25. Specifically when a test spawns a new thread and then calls the library function for loadling the class in the new thread. Here, a class is found, however, when creating a new instance of the object, it is incompatible with the remaining code due to a classloader mismatch apparently.

---

See [mensinda/quarkus-stuff](https://github.com/mensinda/quarkus-stuff/tree/classLoadingError) for the specific implementation.


### Expected behavior

The test also passes with JDK 25

### Actual behavior

```
java.util.concurrent.ExecutionException: java.lang.ClassCastException: class test.TestImpl cannot be cast to class libtest.BaseClass (test.TestImpl is in unnamed module of loader 'app'; libtest.BaseClass is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @41709512)
        at java.base/java.util.concurrent.CompletableFuture.wrapInExecutionException(CompletableFuture.java:345)
        at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:440)
        at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2094)
        at test.LoadingTest.loadingInThreadWithCurrentThreadInLib(LoadingTest.java:111)
        at java.base/java.lang.reflect.Method.invoke(Method.java:565)
        at io.quarkus.test.junit.QuarkusTestExtension.runExtensionMethod(QuarkusTestExtension.java:998)
        at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:846)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
Caused by: java.lang.ClassCastException: class test.TestImpl cannot be cast to class libtest.BaseClass (test.TestImpl is in unnamed module of loader 'app'; libtest.BaseClass is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @41709512)
        at test.LoadingTest.inThreadImplCurrentThread(LoadingTest.java:99)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1825)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1817)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:511)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1450)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2019)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187)
```

### How to Reproduce?

Install both JDK 21 and 25 (and maybe other versions)

[Setup a `toolchains.xml`](https://maven.apache.org/guides/mini/guide-using-toolchains.html)

```bash
git clone git@github.com:mensinda/quarkus-stuff.git
cd quarkus-stuff
git switch classLoadingError

./run.sh 21 # All non-disabled tests pass
./run.sh 25 # One test fails
```

### Output of `uname -a` or `ver`

Linux XXXXX 6.14.0-33-generic #33~24.04.1-Ubuntu SMP PREEMPT_DYNAMIC Fri Sep 19 17:02:30 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "25" 2025-09-16

### Quarkus version or git rev

3.28.4

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.11 (3e54c93a704957b63ee3494413a2b544fd3d825b)

### Additional information

_No response_

--- a/core/deployment/src/main/java/io/quarkus/runner/bootstrap/ForkJoinClassLoading.java
+++ b/core/deployment/src/main/java/io/quarkus/runner/bootstrap/ForkJoinClassLoading.java
@@ -13,11 +13,18 @@ public class ForkJoinClassLoading {
     /**
      * A yucky hack, basically attempt to make sure every thread in the common pool has
      * the correct CL.
-     *
+     * <p>
      * It's not perfect, but as this only affects test and dev mode and not production it is better
      * than nothing.
-     *
+     * <p>
      * Really we should just not use the common pool at all.
+     * <p>
+     * TODO: This no longer works in Java 25 because the `ForkJoinPool` now does
+     * <a href=
+     * "https://github.com/openjdk/jdk/blob/jdk-25%2B36/src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java#L2085">
+     * this</a>, which ends up calling <a href=
+     * "https://github.com/openjdk/jdk/blob/jdk-25%2B36/src/java.base/share/classes/java/util/concurrent/ForkJoinWorkerThread.java#L280">this</a>.
+     * We need to figure out how to deal with this
      */
     public static void setForkJoinClassLoader(ClassLoader classLoader) {
         CountDownLatch allDone = new CountDownLatch(ForkJoinPool.getCommonPoolParallelism());

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/launcher/CustomLauncherInterceptor.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/launcher/CustomLauncherInterceptor.java
@@ -87,6 +87,11 @@ public class CustomLauncherInterceptor implements LauncherDiscoveryListener, Lau
         if (!isProductionModeTests()) {
             initializeFacadeClassLoader();
             adjustContextClassLoader();
+
+            // we need to ensure that the Fork-Join pool will use our thread factory, otherwise the TCCL
+            // of the threads could be wrong
+            System.setProperty("java.util.concurrent.ForkJoinPool.common.threadFactory",
+                    "io.quarkus.bootstrap.forkjoin.QuarkusForkJoinWorkerThreadFactory");
         }
 
     }

